File: code/backend/core/src/main/java/games/paths/core/port/auth/SessionPort.java
package games.paths.core.port.auth;

import games.paths.core.model.auth.RefreshedSession;
import games.paths.core.model.auth.TokenInfo;

/**
 * SessionPort - Inbound port for session and token management.
 * Handles token refresh (with rotation), logout, and token validation.
 * Implemented by {@link games.paths.core.service.auth.SessionService}.
 */
public interface SessionPort {

    /**
     * Refreshes an expired or active access token using a valid refresh token.
     * Token rotation: revokes ALL previous tokens for the user, generates a new pair.
     *
     * @param refreshToken the current refresh token
     * @return a new session with fresh access and refresh tokens, or null if invalid
     */
    RefreshedSession refreshToken(String refreshToken);

    /**
     * Revokes a single refresh token (logout from one device/channel).
     *
     * @param refreshToken the refresh token to revoke
     * @return true if the token was found and revoked, false otherwise
     */
    boolean logout(String refreshToken);

    /**
     * Revokes all active tokens for a user (logout from all devices).
     *
     * @param userUuid the user's public UUID
     * @return true if the user was found and tokens revoked, false otherwise
     */
    boolean revokeAllSessions(String userUuid);

    /**
     * Validates an access token and returns the parsed claims.
     *
     * @param accessToken the JWT access token string
     * @return parsed TokenInfo if valid, null if expired/invalid/malformed
     */
    TokenInfo validateAccessToken(String accessToken);
}

File: code/backend/php/src/Core/Port/Auth/SessionPort.php
<?php

namespace Games\Paths\Core\Port\Auth;

use Games\Paths\Core\Domain\Auth\RefreshedSession;
use Games\Paths\Core\Domain\Auth\TokenInfo;

/**
 * Inbound port for session and token management.
 */
interface SessionPort
{
    /**
     * Refreshes a session using a refresh token.
     * Implements token rotation: revokes all previous tokens for the user.
     */
    public function refreshSession(string $refreshToken): RefreshedSession;

    /**
     * Revokes a specific refresh token (single logout).
     */
    public function logout(string $refreshToken): void;

    /**
     * Revokes all active refresh tokens for a user.
     */
    public function logoutAll(string $userUuid): void;

    /**
     * Validates an access token and returns its claims.
     */
    public function validateAccessToken(string $accessToken): TokenInfo;
}
